File: projections/blazor/ui/hlp.ui.view-runtime/ViewRuntimeTypes.cs
namespace Harborline.UIAdapters.Blazor.Components.DataDisplay;

public sealed record ViewDefinitionField(string Id, string? Label = null);
public sealed record ViewRenderPlanParameters(IReadOnlyList<ViewDefinitionField>? Fields = null);
public sealed record ViewRuntimeAction(string Id, string Label);
public sealed record ViewRenderPlanBindings(string? ViewKind = null, ViewRenderPlanParameters? Parameters = null, IReadOnlyList<ViewRuntimeAction>? Actions = null);
public sealed record ViewRenderPlan(
    string DefinitionHash,
    string DefinitionId,
    string DefinitionVersion,
    string PackKey,
    string PackVersion,
    string DefinitionKind,
    ViewRenderPlanBindings Bindings);
public sealed record ViewRuntimeRow(string Id, IReadOnlyDictionary<string, object?> Values);

File: projections/react/ui/hlp.ui.view-runtime/src/ViewRuntime.types.ts
export const VIEW_KIND_TABLE = 'layout.table' as const

export interface ViewDefinitionField { readonly id: string; readonly label?: string }
export interface ViewRenderPlanParameters { readonly fields?: readonly ViewDefinitionField[] }
export interface ViewRuntimeAction { readonly id: string; readonly label: string }
export interface ViewRenderPlanBindings { readonly viewKind?: string; readonly parameters?: ViewRenderPlanParameters; readonly actions?: readonly ViewRuntimeAction[] }
export interface ViewRenderPlan {
  readonly definitionHash: string
  readonly definitionId: string
  readonly definitionVersion: string
  readonly packKey: string
  readonly packVersion: string
  readonly definitionKind: string
  readonly bindings: ViewRenderPlanBindings
}
export interface ViewRuntimeRow { readonly id: string; readonly [field: string]: unknown }
export interface ViewRuntimeProps {
  readonly plan: ViewRenderPlan
  readonly rows: readonly ViewRuntimeRow[]
  readonly accessibleName?: string
  readonly empty?: string
  readonly actionsDisabled?: boolean
  readonly onRowActivate?: (rowId: string) => void
  readonly onAction?: (actionId: string) => void
}
